pennylane/devices/experimental/default_qubit_2.py

```diff
@@ -283,6 +284,270 @@ def compute_derivatives(
             f"{self.name} cannot compute derivatives via {execution_config.gradient_method}"
         )
 
+    def execute_and_compute_derivatives(
+        self,
+        circuits: QuantumTape_or_Batch,
+        execution_config: ExecutionConfig = DefaultExecutionConfig,
+    ):
+        is_single_circuit = False
+        if isinstance(circuits, QuantumScript):
+            is_single_circuit = True
+            circuits = [circuits]
+
+        if self.tracker.active:
+            for c in circuits:
+                self.tracker.update(resources=c.specs["resources"])
+            self.tracker.update(batches=1, executions=len(circuits))
+            self.tracker.update(derivative_batches=1, derivatives=len(circuits))
+            self.tracker.record()
+
+        if execution_config.gradient_method != "adjoint":
```

Comment:
can this be worked into the `supports_derivatives` method somehow? same for the below functions and their supports_* counterparts

Reply:
would it not be `if execution_config.grad_on_execution and execution_config.use_device_gradient and execution_config.gradient_method != "adjoint"`?

Reply:
Oh wait, is the device itself responsible for checking if gradients need to be computed on the device? I assumed this dispatch was going to be done in the interfaces' `execute` function somewhere.

Reply:
my comment is based on [this](https://github.com/PennyLaneAI/pennylane/blob/9c19af26707ee77d467d2cb24f2531534923173d/pennylane/interfaces/execution.py#L529-L542)

Reply:
Ah I see, ok that makes sense. Yah I can do those extra validation checks here too 👍

Reply:
For some explanation, we want to allow the device to support any type of derivative.  For example, the AWS plugin can natively support parameter shifts. We will need to tidy up the execution config to get that to work, but it will look something like
```
>>> config = ExecutionConfig(gradient_method="parameter-shift", use_device_gradient = None)
>>> dev_aws.preprocess(circuit, config)
ExecutionConfig(gradient_method="parameter-shift", use_device_gradient=True)
```

Reply:
to circle back - can we not do the validation in advance, outside of this function? does any validation depend on run-time stuff?

Reply:
I'd say it's a good idea to do extra validation inside this function even when the dispatch logic is being done elsewhere.

Reply:
I don't think I follow, what cryptic errors do you anticipate? the pre-processing should stop any circuits+devices from getting to this code unless they're fit for execution. if we start mixing things up, we'll lose the value of pre-processing (failing early, cleaner execution code)

Reply:
My worry is that `execute_and_compute_derivatives` will be called from somewhere outside the execution pipeline without having done the pre-processing. For example, if `execute_and_compute_derivatives` is called on a circuit with a non-expval measurement, then `adjoint_jacobian` will fail trying to call `tape.observables`.

Is this something we need to consider?

Reply:
my understanding was that any function with an associated pre-processing/validation should assume that its inputs are already pre-processed/validated. That would mean that if you call it with poorly formed inputs, you will get a cryptic error. However, I don't think the intention is for users to call these things directly, but rather device code will provide implementations of these for the execution pipeline to call (and that pipeline will properly prepare inputs). Is this correct @albi3ro?

Reply:
Alright, I'll remove the extra validation inside the functions then 👍

Reply:
I decided to go ahead and clarify this in the device API:
https://github.com/PennyLaneAI/pennylane/pull/4391

Reply:
if the validation is necessary, it should still happen, just not here. from my above comment (`if execution_config.grad_on_execution and execution_config.use_device_gradient and execution_config.gradient_method != "adjoint"`), I'm proposing that you add that to `supports_derivatives` for DefaultQubit2 so it happens before execution even makes it this far.

Reply:
I understand `use_device_gradient`, but why is the validation for `grad_on_execution` necessary for `supports_derivatives`?

Reply:
`config.grad_on_execution` is just telling the workflow whether or not to call `compute_derivatives` or `execute_and_compute_derivatives`.  So after `preprocessing`, the device doesn't need to pay attention to it.

Reply:
Ah ok, so this doesn't need to be checked here 👍

Reply:
Currently in the QNode when we validate the diff method, we create a new execution config instead of using one returned by `preprocess`:
https://github.com/PennyLaneAI/pennylane/blob/34688855c438a29114bb2df95e4ddab79108fc1b/pennylane/qnode.py#L764-L768
For `DefaultQubit2`, this means that `supports_derivatives` returns `False` here, since it's expecting that `config.use_device_gradient` here is `True`, but is actually `None`. 

What do we want to do in this scenario? This validation happens before we construct the tape, so we can't call `preprocess` yet.

Reply:
```
>>> dev = qml.devices.experimental.DefaultQubit2()
>>> config = qml.devices.experimental.ExecutionConfig(gradient_method="adjoint")
>>> dev.supports_derivatives(config)
True
```

I'd eventually like to move to a better workflow, where things like the gradient method is always chosen based on what we are passing in, but that will take some time, planning, and breaking changes to QNode. This is kind of a patch to mimic the previous behavior. This is just asking "to first order do we support adjoint", and `supports_derivatives` does provide that information.

Reply:
On this branch, `dev.supports_derivatives(config)` returns `False` with the extra validation that @timmysilv suggested (see the failing tests for details). 

In the short run, does adding `use_device_gradient=True` to the execution config work here? Or would that break other devices in the new API?

Reply:
I'm good with what you've added here.